### PR TITLE
Feature/libfcrepo 58

### PR DESCRIPTION
**Pull request by jgottwig**
LIBFCREPO-58. IP address to security principal filter for Tomcat.
- Create a Tomcat valve that adds an HTTP header to the request based on the IP address of the request

https://issues.umd.edu/browse/LIBFCREPO-58

### CODE REVIEW

**Comment by peichman-umd** [comment]
If I add a single IP address instead of a subnet in the mapping file, the valve throws an exception on every request:

```
java.lang.IllegalArgumentException: Could not parse [192.168.40.1]
    org.apache.commons.net.util.SubnetUtils.calculate(SubnetUtils.java:275)
    org.apache.commons.net.util.SubnetUtils.<init>(SubnetUtils.java:51)
    edu.umd.lib.tomcat.valves.IPAddressMapper.getApprovals(IPAddressMapper.java:126)
    edu.umd.lib.tomcat.valves.IPAddressMapper.invoke(IPAddressMapper.java:208)
    edu.umd.lib.tomcat.valves.OptionalSSLAuthenticator.invoke(OptionalSSLAuthenticator.java:40)
    edu.umd.lib.tomcat.valves.OptionalBasicAuthenticator.invoke(OptionalBasicAuthenticator.java:99)
    org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:98)
    org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:950)
    org.apache.catalina.authenticator.SingleSignOn.invoke(SingleSignOn.java:339)
    org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:408)
    org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1040)
    org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:607)
    org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:313)
    java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    java.lang.Thread.run(Thread.java:745)
```

**Comment by jgottwig** [comment]
I've written something to catch this exception, but is the preferred behavior to instead support non-CIDR notation?

**Comment by peichman-umd** [comment]
One last code style comment. In a couple places you declare and initialize a variable to `null` outside of a block, but then only assign and use that variable within that block. In those cases, it would be better to do the declaration and final assignment inside the block. I am going to mark the cases of this I see via line-by-line comments on this PR.

But otherwise, this is terrific work!

File: src/main/java/edu/umd/lib/tomcat/valves/IPAddressMapper.java [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    if (userIP != null && isValidIP(userIP)) {
+      /**
+       * Compare user IP to properties IPs
+       */
+      List<String> approvals = getApprovals(userIP);
+
+      /**
+       * Inject the header with value if the user's IP meets the above criteria.
+       */
+      String finalHeaders = null;
```
Comment (peichman-umd) [inline src/main/java/edu/umd/lib/tomcat/valves/IPAddressMapper.java]: Move this declaration of `finalHeaders` inside the block and combine with its assignment.

File: src/main/java/edu/umd/lib/tomcat/valves/IPAddressMapper.java [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+   * Examine IP and compare against subnets from properties
+   *
+   * @param ip
+   * @return List (approval strings)
+   */
+  protected List<String> getApprovals(String ip) {
+    Enumeration<?> propertyNames = properties.propertyNames();
+
+    List<String> approvals = new ArrayList<String>();
+    SubnetUtils utils; // Our comparison library
```
Comment (peichman-umd) [inline src/main/java/edu/umd/lib/tomcat/valves/IPAddressMapper.java]: Move this declaration of `utils` inside the block and combine with its assignment (on line 132)

**Pull request merged**